File: pwa/src/diary/familyMerge.ts
import type { StoredPlaceState, StoredVisit } from "../catalog/types";
import { nowIso } from "./ids";
import { MAX_PHOTOS_PER_VISIT } from "./photos";
import type { StoredVisitPhoto } from "./types";

export function mergePeople(left: string[], right: string[]): string[] {
  const seen = new Set<string>();
  const names: string[] = [];
  for (const name of [...left, ...right]) {
    const trimmed = name.trim();
    if (!trimmed) {
      continue;
    }
    const key = trimmed.toLocaleLowerCase("cs");
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    names.push(trimmed);
  }
  return names;
}

export function mergeNotes(left: string | null | undefined, right: string | null | undefined): string | null {
  const a = (left || "").trim();
  const b = (right || "").trim();
  if (a && b && a.toLocaleLowerCase("cs") !== b.toLocaleLowerCase("cs")) {
    return `${a} · ${b}`;
  }
  return a || b || null;
}

function richness(visit: StoredVisit, photoCounts: ReadonlyMap<string, number>): number {
  let score = 0;
  if ((visit.note || "").trim()) {
    score += 2;
  }
  if (visit.rating) {
    score += 1;
  }
  score += Math.min(3, visit.people.length);
  score += Math.min(3, photoCounts.get(visit.id) ?? 0);
  return score;
}

export function collapseFamilyVisits(
  visits: StoredVisit[],
  photoCounts: ReadonlyMap<string, number> = new Map(),
): { next: StoredVisit[]; collapsed: number; photoMoves: Array<{ from: string; to: string }> } {
  const groups = new Map<string, StoredVisit[]>();
  for (const visit of visits) {
    if (visit.deleted_at) {
      continue;
    }
    const day = (visit.visited_at || "").trim();
    if (!day) {
      continue;
    }
    const key = `${visit.place_id}|${day}`;
    const rows = groups.get(key) ?? [];
    rows.push(visit);
    groups.set(key, rows);
  }
  const byId = new Map(visits.map((visit) => [visit.id, { ...visit }]));
  let collapsed = 0;
  const photoMoves: Array<{ from: string; to: string }> = [];
  const stamp = nowIso();
  for (const rows of groups.values()) {
    if (rows.length < 2) {
      continue;
    }
    const ranked = [...rows].sort(
      (a, b) =>
        richness(b, photoCounts) - richness(a, photoCounts) ||
        b.created_at.localeCompare(a.created_at) ||
        a.id.localeCompare(b.id),
    );
    const winnerId = ranked[0]?.id;
    if (!winnerId) {
      continue;
    }
    const winner = byId.get(winnerId);
    if (!winner) {
      continue;
    }
    for (const loser of ranked.slice(1)) {
      const current = byId.get(loser.id);
      if (!current) {
        continue;
      }
      winner.people = mergePeople(winner.people, current.people);
      winner.note = mergeNotes(winner.note, current.note);
      if (current.rating != null && (winner.rating == null || current.rating > winner.rating)) {
        winner.rating = current.rating;
      }
      if (current.trip_id && !winner.trip_id) {
        winner.trip_id = current.trip_id;
      }
      current.deleted_at = stamp;
      current.updated_at = stamp;
      photoMoves.push({ from: loser.id, to: winnerId });
      collapsed += 1;
    }
    winner.updated_at = stamp;
    byId.set(winnerId, winner);
  }
  return { next: [...byId.values()], collapsed, photoMoves };
}

export function unionFamilyStates(local: StoredPlaceState[], incoming: StoredPlaceState[]): StoredPlaceState[] {
  const byId = new Map(local.map((state) => [state.place_id, { ...state }]));
  const stamp = nowIso();
  for (const item of incoming) {
    const existing = byId.get(item.place_id);
    if (!existing) {
      byId.set(item.place_id, item);
      continue;
    }
    existing.want_to_visit = existing.want_to_visit || item.want_to_visit;
    existing.favorite = existing.favorite || item.favorite;
    existing.personal_note = mergeNotes(existing.personal_note, item.personal_note);
    existing.deleted_at = null;
    existing.updated_at = stamp;
    byId.set(item.place_id, existing);
  }
  return [...byId.values()];
}

export function reassignPhotos(
  photos: StoredVisitPhoto[],
  moves: Array<{ from: string; to: string }>,
): StoredVisitPhoto[] {
  const map = new Map(moves.map((row) => [row.from, row.to]));
  const counts = new Map<string, number>();
  const kept: StoredVisitPhoto[] = [];
  const moved: StoredVisitPhoto[] = [];
  for (const photo of photos) {
    const nextVisit = map.get(photo.visit_id);
    if (nextVisit) {
      moved.push({ ...photo, visit_id: nextVisit });
      continue;
    }
    const n = counts.get(photo.visit_id) ?? 0;
    if (n >= MAX_PHOTOS_PER_VISIT) {
      continue;
    }
    counts.set(photo.visit_id, n + 1);
    kept.push(photo);
  }
  for (const photo of moved) {
    const n = counts.get(photo.visit_id) ?? 0;
    if (n >= MAX_PHOTOS_PER_VISIT) {
      continue;
    }
    counts.set(photo.visit_id, n + 1);
    kept.push(photo);
  }
  return kept;
}

File: pc-app/app/services/family_merge.py
"""Rodinné sloučení deníku: stejné místo a den = jedno razítko."""

from __future__ import annotations

import json
import shutil
from collections import defaultdict

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db.models import PlaceJournalState, Visit, now_iso
from app.services.visit_photos import (
    MAX_PHOTOS_PER_VISIT,
    is_visit_id,
    list_visit_photos,
    visit_photo_dir,
)


def merge_people(left: list[str], right: list[str]) -> list[str]:
    seen: set[str] = set()
    names: list[str] = []
    for name in [*left, *right]:
        trimmed = name.strip()
        if not trimmed:
            continue
        key = trimmed.casefold()
        if key in seen:
            continue
        seen.add(key)
        names.append(trimmed)
    return names


def merge_notes(left: str | None, right: str | None) -> str | None:
    a = (left or "").strip()
    b = (right or "").strip()
    if a and b and a.casefold() != b.casefold():
        return f"{a} · {b}"
    return a or b or None


def visit_richness(visit: Visit) -> tuple[int, str]:
    score = 0
    if visit.note and visit.note.strip():
        score += 2
    if visit.rating:
        score += 1
    score += min(3, len(visit.people))
    photos = list_visit_photos(visit.public_id) if is_visit_id(visit.public_id) else []
    score += min(3, len(photos))
    return (score, visit.created_at)


def _family_rank(visit: Visit) -> tuple:
    """Stejné pořadí jako PWA: vyšší skóre, novější created_at, menší id."""
    score, created = visit_richness(visit)
    return (-score, tuple(-ord(c) for c in created), visit.public_id)


def _move_photos(loser_id: str, winner_id: str) -> None:
    if loser_id == winner_id or not is_visit_id(loser_id) or not is_visit_id(winner_id):
        return
    source = list_visit_photos(loser_id)
    if not source:
        return
    dest = visit_photo_dir(winner_id)
    folder = source[0].parent
    dest_count = len(list_visit_photos(winner_id))
    for path in source:
        target = dest / path.name
        if target.exists() or dest_count >= MAX_PHOTOS_PER_VISIT:
            path.unlink(missing_ok=True)
            continue
        shutil.move(path, target)
        dest_count += 1
    try:
        folder.rmdir()
    except OSError:
        pass


def collapse_family_visits(session: Session) -> int:
    """Stejné place_id + visited_at (s datem) sloučí do jedné živé návštěvy."""
    # Session má autoflush=False — bez flush SELECT neuvidí návštěvy právě přidané importem.
    session.flush()
    live = list(session.scalars(select(Visit).where(Visit.deleted_at.is_(None))).all())
    groups: dict[tuple[str, str], list[Visit]] = defaultdict(list)
    for visit in live:
        day = (visit.visited_at or "").strip()
        if not day:
            continue
        groups[(visit.place_public_id, day)].append(visit)
    collapsed = 0
    stamp = now_iso()
    for rows in groups.values():
        if len(rows) < 2:
            continue
        ranked = sorted(rows, key=_family_rank)
        winner = ranked[0]
        for loser in ranked[1:]:
            winner.people_json = json.dumps(merge_people(winner.people, loser.people), ensure_ascii=False)
            winner.note = merge_notes(winner.note, loser.note)
            if loser.rating is not None and (winner.rating is None or loser.rating > winner.rating):
                winner.rating = loser.rating
            if loser.trip_public_id and not winner.trip_public_id:
                winner.trip_public_id = loser.trip_public_id
            _move_photos(loser.public_id, winner.public_id)
            loser.deleted_at = stamp
            loser.updated_at = stamp
            collapsed += 1
        winner.updated_at = stamp
    return collapsed


def family_or_state(local: PlaceJournalState, incoming: dict) -> None:
    local.want_to_visit = 1 if local.want_to_visit or incoming.get("want_to_visit") else 0
    local.favorite = 1 if local.favorite or incoming.get("favorite") else 0
    local.personal_note = merge_notes(local.personal_note, incoming.get("personal_note"))
    local.updated_at = now_iso()
    local.deleted_at = None
